### PR TITLE
Adding Junit5 Operator upgrade tests

### PR DESCRIPTION
**Pull request by vanajamukkara**
Adding upgrade tests for 
2.5.0 to latest
2.6.0 to latest
3.0.0 to latest

excluding upgrade test from full test as they should not be run in parallel. 
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/944/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-model-in-image-tests-10/598/

### CODE REVIEW

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+    logger.info("Assign a unique namespace for domain");
+    assertNotNull(namespaces.get(2), "Namespace is null");
+    String domainNamespace = namespaces.get(2);
+    latestOperatorImageName = getOperatorImageName();
+
+    // delete existing CRD
+    new Command()
+        .withParams(new CommandParams()
+            .command("kubectl delete crd domains.weblogic.oracle --ignore-not-found"))
+        .execute();
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java:170]: what happens when the test fails after this stage? The subsequent tests fails without a CRD?

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+
+    // install operator
+    String opNamespace = opNamespace1;
+    String opServiceAccount = opNamespace + "-sa";
+    installAndVerifyOperator(opNamespace, opServiceAccount, true,
+        0, opHelmParams, domainNamespace);
+
+    // create domain
+    createDomainHomeInImageAndVerify(domainNamespace, operatorVersion);
+
```
Comment (sankarpn) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java:188]: If the tests fails here the CRD version will not be latest?

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java (hunk 1 above)
Comment (vanajamukkara) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java:170]: The next step in the test is to install Operator which creates CRD. If the test fails after deleting the CRD and before installing the operator, for the next upgrade test the CRD will be created when the operator is installed. subsequent tests will not fail.

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java (hunk 2 above)
Comment (vanajamukkara) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java:188]: If the test fails, CRD is deleted in the afterEach method.

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java (hunk 1 above)
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java:170]: My concern is if any of the upgrade test(may be the last upgradetest) fails and leaves the env with a missing CRD it is going to affect the other tests in other IT classes right?

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java (hunk 2 above)
Comment (sankarpn) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java:188]: ok good.

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+/**
+ * Operator upgrade tests.
+ */
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+@DisplayName("Operator upgrade tests")
+@IntegrationTest
+public class ItOperatorUpgrade {
+
+  public static final String LATEST_API_VERSION = "v8";
+  private static ConditionFactory withStandardRetryPolicy = null;
```
Comment (marinakog) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java:73]: Should we move LATEST_API_VERSION to test constants and make it configurable with env var?

File: new-integration-tests/src/test/resources/domain/domain-260.yaml [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    - name: USER_MEM_ARGS
+      value: "-Djava.security.egd=file:/dev/./urandom "
+    # volumes:
+    # - name: weblogic-domain-storage-volume
+    #   persistentVolumeClaim:
+    #     claimName: domain1-weblogic-sample-pvc
+    # volumeMounts:
+    # - mountPath: /shared
+    #   name: weblogic-domain-storage-volume
+
```
Comment (marinakog) [inline new-integration-tests/src/test/resources/domain/domain-260.yaml:70]: remove commented out pvc config

File: new-integration-tests/src/test/resources/domain/domain-260.yaml [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    # serverStartState legal values are "RUNNING" or "ADMIN"
+    # "RUNNING" means the listed server will be started up to "RUNNING" mode
+    # "ADMIN" means the listed server will be start up to "ADMIN" mode
+    serverStartState: "RUNNING"
+    adminService:
+      channels:
+    # The Admin Server's NodePort
+       - channelName: default
+         nodePort: 30802
+    # Uncomment to export the T3Channel as a service
```
Comment (marinakog) [inline new-integration-tests/src/test/resources/domain/domain-260.yaml:82]: hardcoded nodePort might effect runs in parallel

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java (hunk 1 above)
Comment (vanajamukkara) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java:170]: The idea is upgrade tests should delete the CRD(as it uses old operator releases) after the testing is done(success or failed). All the IT classes install the operator first, this is when CRD is created hence there won't be a situation where the tests will be run without CRD.

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java (hunk 3 above)
Comment (vanajamukkara) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorUpgrade.java:73]: Done. Also using existing constant.

File: new-integration-tests/src/test/resources/domain/domain-260.yaml (hunk 4 above)
Comment (vanajamukkara) [reply new-integration-tests/src/test/resources/domain/domain-260.yaml:63]: Done

File: new-integration-tests/src/test/resources/domain/domain-260.yaml (hunk 5 above)
Comment (vanajamukkara) [reply new-integration-tests/src/test/resources/domain/domain-260.yaml:75]: Upgrade tests should not be run in parallel with other tests, it is excluded while running all the tests in parallel. But it is better not to hardcode the nodeport, changed to 0 to pick the available port.

**Comment by vanajamukkara** [comment]
Jenkins results with changes -
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/972/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-model-in-image-tests-10/619/

**Pull request merged**